#Phase1:

##Goal of Collecting Dataset:

The primary goal of analyzing this dataset is to:

Performance Analysis: Assess team and player performances by examining various metrics such as goals, assists, and other key statistics.

Outcome Prediction: Develop models to predict match outcomes or player performances based on historical data.

Trend Identification: Discover trends and patterns in match results, player statistics, and team strategies.

We will achieve this by clustering data and classifying it using machine learning techniques acquired during the course.

##Source of Dataset:
Kaggle Platform Dataset link: https://www.kaggle.com/datasets/evangower/fifa-world-cup

##General Information about the Dataset:

Number of Attributes:15

Number of Objects:900 (Match)

Class Name/Labels: outcome


In [2]:

import pandas as pd

df = pd.read_csv('Dataset/wcmatches.csv')

print(df.info())

num_row = len(df)
print("Number of rows:", num_row)


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 900 entries, 0 to 899
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   year            900 non-null    int64 
 1   country         900 non-null    object
 2   city            900 non-null    object
 3   stage           900 non-null    object
 4   home_team       900 non-null    object
 5   away_team       900 non-null    object
 6   home_score      900 non-null    int64 
 7   away_score      900 non-null    int64 
 8   outcome         900 non-null    object
 9   win_conditions  62 non-null     object
 10  winning_team    731 non-null    object
 11  losing_team     731 non-null    object
 12  date            900 non-null    object
 13  month           900 non-null    object
 14  dayofweek       900 non-null    object
dtypes: int64(3), object(12)
memory usage: 105.6+ KB
None
Number of rows: 900


------------------------------------------------------

#Phase2:

| Attributes Name | Data type | Description | Possible Values |
|----------|----------|----------|----------|
|Year |	Integer| The year of the match |	Any year (e.g., 1930, 1934, etc.)|
|Country |	String |The country hosting the World Cup |  Categorical string values (different possible)|
|City	| String |	The city where the match took place	 | Categorical string values (different possible)|
|Stage	| String|	The stage of the tournament | Categorical string values (different possible)|
|Home Team |	String | The team playing at home |	Categorical string values (different possible)|
|Away Team |	String |	The team playing away |	Categorical string values (different possible)|
|Home Score |	Integer	| The score of the home team |	0-10|
|Away Score	| Integer |	The score of the away team |	0-8|
|Outcome |	String | The result of the match |	H (Home win), A (Away win), D (Draw)|
|Win Conditions |	String |	Conditions under which the winning team was determined	Specific phrases |Categorical string values (different possible)|
|Winning Team |	String | The team that won the match |	Categorical string values (different possible)|
|Losing Team|	String | The team that lost the match |	Categorical string values (different possible)|
|Date |Date |	The date of the match |	Datetime |
|Month |String	|The month in which the match took place|Categorical string values (different possible)|
|DayOfWeek |	String |	The day of the week when the match occurred | Categorical string values (different possible)|


##Sample

In [3]:

sample = df.sample(frac=1)
print(sample);

     year      country           city          stage       home_team  \
883  2018       Russia    Kaliningrad        Group G         England   
224  1970       Mexico    Guadalajara  Quarterfinals          Brazil   
58   1950       Brazil      São Paulo        Group 1          Brazil   
334  1982        Spain         Málaga        Group 6    Soviet Union   
449  1990        Italy           Bari    Round of 16  Czechoslovakia   
..    ...          ...            ...            ...             ...   
872  2018       Russia          Sochi        Group C       Australia   
521  1998       France    Montpellier        Group D        Paraguay   
290  1978    Argentina  Mar del Plata        Group 3         Austria   
82   1954  Switzerland       Lausanne        Group 4     Switzerland   
378  1986       Mexico    Mexico City        Group B          Mexico   

       away_team  home_score  away_score outcome win_conditions  \
883      Belgium           0           1       A            NaN   
2

##Show the missing value:

In [4]:
missing_values = df.isnull().sum()
print("Missing values ")
print(missing_values )
missing_counts = missing_values.sum();
print("The total of Missing value is :")
print(missing_counts)

Missing values 
year                0
country             0
city                0
stage               0
home_team           0
away_team           0
home_score          0
away_score          0
outcome             0
win_conditions    838
winning_team      169
losing_team       169
date                0
month               0
dayofweek           0
dtype: int64
The total of Missing value is :
1176


##Show the Min., 1st Qu., Median, Mean ,3rd Qu.,Max. for each numeric column:

In [5]:
summary_stats=df.describe();
print(summary_stats);

              year  home_score  away_score
count   900.000000  900.000000  900.000000
mean   1986.915556    1.568889    1.262222
std      23.150270    1.489574    1.313960
min    1930.000000    0.000000    0.000000
25%    1970.000000    0.000000    0.000000
50%    1990.000000    1.000000    1.000000
75%    2006.000000    2.000000    2.000000
max    2018.000000   10.000000    8.000000


##Show the Variance:

In [6]:
variance= df.var(numeric_only=True);
print(variance);

year          535.935019
home_score      2.218831
away_score      1.726491
dtype: float64


##Data Preprocessing:

###Data Cleaning:

###1-Remove duplicates:

####First, we count the number of duplicates rows, and then we remove them:

In [7]:
import pandas as pd
data = pd.read_csv('Dataset/wcmatches.csv')
num_duplicates = data.duplicated().sum()
data_cleaned = data.drop_duplicates()
print("Number of duplicate rows:", num_duplicates)
print("DataFrame after dropping all duplicate rows:")
print(data_cleaned)
data_cleaned.to_csv('Cleaned_dataset.csv', index=False)

Number of duplicate rows: 0
DataFrame after dropping all duplicate rows:
     year  country              city          stage  home_team      away_team  \
0    1930  Uruguay        Montevideo        Group 1     France         Mexico   
1    1930  Uruguay        Montevideo        Group 4    Belgium  United States   
2    1930  Uruguay        Montevideo        Group 2     Brazil     Yugoslavia   
3    1930  Uruguay        Montevideo        Group 3       Peru        Romania   
4    1930  Uruguay        Montevideo        Group 1  Argentina         France   
..    ...      ...               ...            ...        ...            ...   
895  2018   Russia             Sochi  Quarterfinals     Russia        Croatia   
896  2018   Russia  Saint Petersburg     Semifinals     France        Belgium   
897  2018   Russia            Moscow     Semifinals    Croatia        England   
898  2018   Russia  Saint Petersburg    Third place    Belgium        England   
899  2018   Russia            Moscow

This shows that our dataset has no duplicate rows, so the total number of rows remains unchanged at 900.

###2-Handling Missing Values: 

####First, we calculate the number of missing values in our dataset:

In [8]:
import pandas as pd
data = pd.read_csv('Cleaned_dataset.csv');
sample=data.sample(n=20);
print(sample);
missing_values = data.isna()
missing_counts = missing_values.sum();
rows_with_missing = data[data.isna().any(axis=1)];

print("Missing values in each column:");
print(missing_counts);
print("\nRows with missing values:");
print(rows_with_missing);

     year        country               city        stage     home_team  \
483  1994  United States            Pontiac      Group B        Sweden   
710  2010   South Africa       Johannesburg      Group B     Argentina   
73   1950         Brazil     Rio de Janeiro  Final Round        Brazil   
435  1990          Italy              Udine      Group E   South Korea   
13   1930        Uruguay         Montevideo      Group 3       Uruguay   
265  1974        Germany  Frankfurt am Main      Group B  West Germany   
24   1934          Italy               Rome  Round of 16         Italy   
661  2006        Germany            Hamburg      Group A       Ecuador   
438  1990          Italy           Florence      Group A       Austria   
25   1934          Italy              Milan  Round of 16   Netherlands   
271  1978      Argentina       Buenos Aires      Group 1     Argentina   
736  2010   South Africa       Bloemfontein      Group F      Slovakia   
354  1982          Spain          Barc

Missing values in each column:
year                0
country             0
city                0
stage               0
home_team           0
away_team           0
home_score          0
away_score          0
outcome             0
win_conditions    838
winning_team      169
losing_team       169
date                0
month               0
dayofweek           0
dtype: int64

Rows with missing values:
     year  country              city          stage  home_team      away_team  \
0    1930  Uruguay        Montevideo        Group 1     France         Mexico   
1    1930  Uruguay        Montevideo        Group 4    Belgium  United States   
2    1930  Uruguay        Montevideo        Group 2     Brazil     Yugoslavia   
3    1930  Uruguay        Montevideo        Group 3       Peru        Romania   
4    1930  Uruguay        Montevideo        Group 1  Argentina         France   
..    ...      ...               ...            ...        ...            ...   
893  2018   Russia             K

This shows that our dataset has missing values in multiple columns: 838 in the 'win_conditions' column, 169 in the 'winning_team' column, and 169 in the 'losing_team' column.

####Secondly, we handle these missing values by calculating the mode value for the target columns ['win_conditions', 'winning_team', 'losing_team'], and then we replace the missing values with the mode.

In [9]:
import pandas as pd

data = pd.read_csv('Cleaned_dataset.csv')

data.columns = data.columns.str.strip()


for column in ['win_conditions', 'winning_team', 'losing_team']:
    modal_value = data[column].mode()[0]  
    data[column] = data[column].fillna(value=modal_value)  

print("Missing values after filling:")
print(data.isnull().sum())

data.to_csv('Cleaned_dataset.csv', index=False)



Missing values after filling:
year              0
country           0
city              0
stage             0
home_team         0
away_team         0
home_score        0
away_score        0
outcome           0
win_conditions    0
winning_team      0
losing_team       0
date              0
month             0
dayofweek         0
dtype: int64


and Here, the output shows that we handled all the missing values.

###3-Handling outliers:

Before we handle our outliers we did choose the IQR method Based on our dataset to identify how many outliers in our dataset.

In [10]:
import pandas as pd
data = pd.read_csv('Cleaned_dataset.csv')
import numpy as np
outlier_threshold = 1.5

def count_outliers(column_data):
    q1 = np.percentile(column_data, 25)
    q3 = np.percentile(column_data, 75)
    iqr = q3 - q1
    upper_bound = q3 + outlier_threshold * iqr
    lower_bound = q1 - outlier_threshold * iqr
    outliers = (column_data > upper_bound) | (column_data < lower_bound)
    return sum(outliers)

numeric_columns = data.select_dtypes(include=[np.number]).columns

outlier_counts = {}
total_rows_with_outliers = 0

for column in numeric_columns:
    outliers = count_outliers(data[column])
    outlier_counts[column] = outliers
    total_rows_with_outliers += outliers

total_rows = len(data)

print("Outlier Counts:")
for column, count in outlier_counts.items():
    print(f"{column}: {count} rows with outliers")

print(f"Total Rows with Outliers: {total_rows_with_outliers}")

Outlier Counts:
year: 0 rows with outliers
home_score: 24 rows with outliers
away_score: 11 rows with outliers
Total Rows with Outliers: 35


The analysis of our dataset revealed that a total of 35 rows contain outliers, which is a small proportion compared to the overall dataset of 900 rows. Specifically, there are 0 rows with outliers in the 'year' column, 24 rows in the 'home_score' column, and 11 rows in the 'away_score' column. Although this represents a limited number of outliers, their presence indicates some extreme values that deviate from the majority of the data. To address this issue, we have decided to adopt a strategy of capping the outliers. Rather than removing these data points entirely, we opted to replace them with the nearest non-outlier values. This approach allows us to retain the entirety of our dataset while mitigating the impact of these extreme values on our subsequent analyses. By capping the outliers and maintaining the remaining data, we aim to preserve the integrity of the dataset and avoid the potential loss of valuable information that would occur through outright removal of the outliers. This ensures that our analysis incorporates as much relevant information as possible while still accounting for the presence of extreme observations

####handling outliers:

####Checking the results by counting outliers after handling them.